### PR TITLE
SNOW-624882: flatten df join table function

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-624882

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   This PR addresses `dataframe.join_table_function` and parts of `dataframe.select` as a larger effort to simplify the sql query generated on the client side and reduce the amount of nesting.

### CODE REVIEW

File: src/snowflake/snowpark/dataframe.py [hunk 1]
```diff
@@ -778,11 +779,20 @@ def select(
...
+                            join_plan, analyzer=self._session._analyzer
+                        ),
+                        analyzer=self._session._analyzer,
+                    ).select(names)
+                )
             return self._with_plan(self._select_statement.select(names))
         else:
+            if join_plan:
+                return self._with_plan(Project(names, join_plan))
             return self._with_plan(Project(names, self._plan))
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe.py:796]: How about `return self._with_plan(Project(names, join_plan or self._plan))`?

File: src/snowflake/snowpark/dataframe.py [hunk 2]
```diff
@@ -1872,15 +1882,42 @@ def join_table_function(
...
+            return self._with_plan(
+                SelectStatement(
+                    from_=SelectTableFunction(
+                        func_expr,
+                        other_plan=self._plan,
+                        analyzer=self._session._analyzer,
+                    ),
+                    analyzer=self._session._analyzer,
+                )
+            )
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe.py:1919]: I'm wondering if the following makes sense:

```
names = None
if func_expr.aliases:
  ...
  names = [*old_cols, *new_cols]

if context._use_sql_simplifier:
  select_plan = self._with_plan(...)
  if names:
    select_plan = select_plan.select(names)
  return self._with_plan(select_plan)

if names:
  return self._with_plan(Project(names, join_plan))
return self._with_plan(TableFunctionJoin(self._plan, func_expr))
```

Main goal is to reduce the duplication of `SelectStatement`

File: src/snowflake/snowpark/dataframe.py (hunk 1 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/dataframe.py:796]: good idea

File: tests/integ/test_simplifier_suite.py [hunk 3]
```diff
@@ -331,6 +333,130 @@ def test_select_expr(session, simplifier_table):
...
+    @udtf(output_schema=["two_x", "three_x"])
+    class multiplier_udtf:
+        def process(self, n: int) -> Iterable[Tuple[int, int]]:
+            yield (2 * n, 3 * n)
+
+    df1 = df.select("a", split_to_table("c", lit(" ")))
+    df2 = df.select("a", multiplier_udtf(df.b))
+    # test multiple selects are flattened
+    df3 = df1.select("a", "seq").select("a")
+    assert df1.queries["queries"][-1].count("SELECT") == df3.queries["queries"][
```
Comment (sfc-gh-jdu) [inline tests/integ/test_simplifier_suite.py:352]: You should assert that the result equals a value?

Comment (sfc-gh-aalam) [reply tests/integ/test_simplifier_suite.py:352]: I am adding output equals checks for repeated select. rest are tested in regular tests

**Review by sfc-gh-jdu — APPROVED** [review]

File: tests/integ/test_simplifier_suite.py (hunk 3 above)
Comment (sfc-gh-jdu) [reply tests/integ/test_simplifier_suite.py:354]: I see

**Pull request merged**